internal/pkg/vm/internal/vmcontext/vmcontex2.go

```diff
@@ -0,0 +1,543 @@
+	// flush all objects out
+	if err := vm.store.Flush(); err != nil {
+		// Review: what do we want to do with this errors?
+		//         we can change the return to be ([]Receipt, error) although its unclear what to do about it as a caller
+		//         we can just crash, or let the caller retry a few times and then crash
+		panic(err)
+	}
+	// commit new actor state
+	if err := vm.state.Commit(context.Background()); err != nil {
+		// Review: what do we want to do with this errors?
+		//         similarly, what can the caller do?
+		panic(err)
+	}
+	// TODO: update state root (issue: #3718)
+
+	return receipts
+}
+
+// applyImplicitMessage applies messages automatically generated by the vm itself.
+//
+// This messages do not consume client gas are do not fail.
+func (vm *VM) applyImplicitMessage(imsg internalMessage) interface{} {
+	// implicit messages gas is tracked separatly and not paid by the miner
+	gasTank := gas.NewTracker(gas.SystemGasLimit)
+
+	// the execution of the implicit messages is simpler than full external/actor-actor messages
+	// execution:
+	// 1. load from actor
+	// 2. increment seqnumber
+	// 3. build new context
+	// 4. invoke message
+
+	// 1. load from actor
+	fromActor, err := vm.state.GetActor(context.Background(), imsg.from)
+	if fromActor == nil || err != nil {
+		panic("implicit message `from` actor not found")
+	}
+
+	// 2. increment seq number
+	fromActor.IncrementSeqNum()
+
+	// 3. build context
+	ctx := newInvocationContext(vm, imsg, fromActor, &gasTank)
+
+	// 4. invoke message
+	// Note: any errors raised by the invocation will not be trapped
+	// 		 errors are not expected by the spec on implicit messages, this messages must work
+	return ctx.invoke()
+}
+
+// applyMessage applies the message to the current state.
+func (vm *VM) applyMessage(msg *types.UnsignedMessage, onChainMsgSize uint32, miner address.Address) (message.Receipt, MinerPenaltyFIL) {
+	// This method does not actually execute the message itself,
+	// but rather deals with the pre/post processing of a message.
+	// (see: `invocationContext.invoke()` for the dispatch and execution)
+
+	// initiate gas tracking
+	gasTank := gas.NewTracker(msg.GasLimit)
+
+	// pre-send
+	// 1. charge for message existence
+	// 2. load sender actor
+	// 3. check message seq number
+	// 4. check if _sender_ has enough funds
+	// 5. increment message seq number
+	// 6. withheld maximum gas from _sender_
+	// 7. checkpoint state
+
+	// 1. charge for bytes used in chain
+	ok := gasTank.TryCharge(gascost.OnChainMessage(onChainMsgSize))
+	if !ok {
+		// Invalid message; insufficient gas limit to pay for the on-chain message size.
+		return message.Failure(exitcode.OutOfGas), gasTank.GasConsumed().Cost(msg.GasPrice)
+	}
+
+	// Dragons: from address need to be normalized here
+
+	// 2. load actor from global state
+	fromActor, err := vm.state.GetActor(context.Background(), msg.From)
+	// Review: when are errors surfaced? this can lead to invalid state transitions
```

Comment:
What do you mean by invalid? Differing from the spec / other impls?

Reply:
Yea, invalid as in differ from other impls. 

That `GetActor` could be possible hitting disk, in which case an IO error in that path by other implementations might not be surfaced as a `ActorNotFound` exitcode which will result in a receipt being generated.

Reply:
Right, I think we need to distinguish those error here. How about returning `(actor, ok, err)`, so errors are only for unrecoverables?